validation/validate_metadata.py

```diff
@@ -0,0 +1,446 @@
+#! /usr/bin/python
+"""Validate input metadata tsv file against metadata convention.
+
+DESCRIPTION
+This CLI takes a tsv metadata file and validates against a metadata convention
+using the python jsonschema library. The metadata convention JSON schema
+represents the rules that should be enforced on metadata files for studies
+participating under the convention.
+
+EXAMPLE
+# Using json file for Alexandria metadata convention tsv, validate input tsv
+$ validate_metadata.py AMC_v0.8.json metadata_test.tsv
+
+"""
+
+import argparse
+import json
+import jsonschema
+from collections import defaultdict
+import logging
+import ast
+from itertools import islice
+
+# ToDo set up parameters to adjust log levels
+#  logging.basicConfig(level=logging.INFO, filename='app.log', filemode='w',
+#   format='%(name)s - %(levelname)s - %(message)s')
+logging.basicConfig(
+    level=logging.INFO, format='%(name)s - %(levelname)s - %(message)s'
+)
+logger = logging.getLogger(__name__)
+
+
+class Cell_Metadata:
+    """A class used to represent cell metadata.
+
+    Attributes
+    ----------
+    headers : list of strings
+        list of metadata names in metadata file
+    metadata_types : list of strings
+        list of metadata type designations from metadata file
+    annotation_type : list of strings
+        list of valid metadata type designations
+    errors : dict
+        dict of collected validation errors and warnings
+
+    """
+    def __init__(self, file_path):
+        """
+        Parameters
+        ----------
+        filepath : str
+            The name of the metadata file
+
+        """
+        self.file = open(file_path, 'r')
+        self.headers = self.file.readline().rstrip('\n').split('\t')
+        self.metadata_types = self.file.readline().rstrip('\n').split('\t')
+        self.annotation_type = ['group', 'numeric']
+        self.errors = defaultdict(list)
+        self.ontology = defaultdict(lambda: defaultdict(set))
+        self.type = defaultdict(list)
+
+    def validate_header_keyword(self):
+        """Check metadata header row starts with NAME (case-insensitive).
+
+        :return: boolean   True if valid, False otherwise
+        """
+        logger.debug('Begin: validate_header_keyword')
+        valid = False
+        if self.headers[0].casefold() == 'NAME'.casefold():
+            valid = True
+            if self.headers[0] != 'NAME':
+                # ToDO - capture warning below in error report
+                logger.warning(
+                    'Warning: metadata file keyword "NAME" provided as {x}'.
+                    format(x=self.headers[0])
+                )
+        else:
+            self.errors['format'].append(
+                ('Error: Metadata file header row malformed, missing NAME')
```

Comment:
testing removal extraneous parens - the autoformatter did something odd because of the length of the line. Intending to switch to black to see what it will do for this case but having yet had time to try it.